File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        count = 0
        def countGood(root, stack):
            if root is None:
                return
            
            print(f'val is {root.val} and stack {stack}')

            nonlocal count
            if len(stack) == 0 or root.val >= stack[-1]:
                stack.append(root.val)
                count += 1
            
            countGood(root.left, stack[:])
            countGood(root.right, stack[:])
        countGood(root, [])
        return count
            

            
            

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        // we have three cases a->c<-b, a->b, b<-a
        // anyhow its just the case that 
        if (p->val >= q->val) swap(p, q);
        if (!root) return root;
        if (root->val >= p->val && root->val <= q->val) return root;
        auto leftRet{lowestCommonAncestor(root->left, p, q)};
        if (leftRet) return leftRet;
        return lowestCommonAncestor(root->right, p, q);
    }
};
